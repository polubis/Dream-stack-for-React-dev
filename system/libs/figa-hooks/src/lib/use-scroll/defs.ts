import type { MutableRefObject } from 'react';

interface BaseScrollState {
  prev: number;
  curr: number;
}

interface IdleScrollState {
  is: 'idle';
}

interface ProgressScrollState extends BaseScrollState {
  is: 'progress';
}

interface RegressScrollState extends BaseScrollState {
  is: 'regress';
}

interface UnchangedScrollState extends BaseScrollState {
  is: 'unchanged';
}

type ScrollAxis = 'x' | 'y';

type ScrollState =
  | IdleScrollState
  | ProgressScrollState
  | RegressScrollState
  | UnchangedScrollState;

type ScrollResult = ScrollState['is'];

interface ScrollConfig {
  axis?: ScrollAxis;
  delay?: number;
}

type ScrollReturn<T extends HTMLElement = HTMLElement> = Readonly<
  [ScrollState, MutableRefObject<T | null>]
>;

export type {
  IdleScrollState,
  ProgressScrollState,
  RegressScrollState,
  ScrollAxis,
  UnchangedScrollState,
  BaseScrollState,
  ScrollReturn,
  ScrollConfig,
  ScrollResult,
  ScrollState,
};
