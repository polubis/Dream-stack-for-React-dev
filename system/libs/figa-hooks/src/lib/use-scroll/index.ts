export * from './use-scroll';
export * from './defs';
