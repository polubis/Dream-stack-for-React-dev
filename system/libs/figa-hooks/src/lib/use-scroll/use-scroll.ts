import { useRef, useState } from 'react';
import { useIsomorphicLayoutEffect } from '../use-isomorphic-layout-effect';
import type {
  ScrollAxis,
  ScrollConfig,
  ScrollReturn,
  ScrollResult,
  ScrollState,
} from './defs';
import { debounceTime, fromEvent } from 'rxjs';

const warn = (): void => {
  console.warn(`
      Can't find the reference, and the window object. So this hook will not work.
  
      It may be caused because: 
        - you're rendering ref conditionally,
        - you forgot to use forwardRef() - when ref passed to component,
        - you're using this hook on a server.
    `);
};

const isHTMLElement = <T extends HTMLElement>(ref: T | Window): ref is T =>
  (ref as T).offsetTop !== undefined;

const getScrollYValue = <T extends HTMLElement>(ref: T | Window): number =>
  isHTMLElement(ref) ? ref.scrollTop : ref.scrollY;

const getScrollXValue = <T extends HTMLElement>(ref: T | Window): number =>
  isHTMLElement(ref) ? ref.scrollLeft : ref.scrollX;

const getScrollValue = <T extends HTMLElement>(
  axis: ScrollAxis,
  ref: T | Window
): number => (axis === 'x' ? getScrollXValue(ref) : getScrollYValue(ref));

const getScrollResult = (prev: number, curr: number): ScrollResult => {
  if (prev === curr) return 'unchanged';
  return prev < curr ? 'progress' : 'regress';
};

const useScroll = <T extends HTMLElement = HTMLElement>({
  axis = 'y',
  delay = 150,
}: ScrollConfig = {}): ScrollReturn<T> => {
  const ref = useRef<T>(null);
  const [state, setState] = useState<ScrollState>(() => ({ is: 'idle' }));

  useIsomorphicLayoutEffect(() => {
    const target = ref.current ? ref.current : window;

    if (!target) {
      warn();
      return;
    }

    let prev = getScrollValue(axis, target);

    const obs$ = fromEvent(target, 'scroll')
      .pipe(debounceTime(delay))
      .subscribe(() => {
        const curr = getScrollValue(axis, target);

        setState({
          prev,
          curr,
          is: getScrollResult(prev, curr),
        });

        prev = curr;
      });

    return () => {
      obs$.unsubscribe();
    };
  }, [delay, axis]);

  return [state, ref];
};

export { useScroll };
