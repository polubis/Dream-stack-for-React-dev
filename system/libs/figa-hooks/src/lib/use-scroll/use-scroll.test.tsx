import {
  fireEvent,
  render,
  renderHook,
  screen,
  waitFor,
} from '@testing-library/react';
import { useScroll } from './use-scroll';
import type { ScrollAxis, ScrollConfig, ScrollResult } from './defs';

describe('Scroll data detection works when:', () => {
  const getFixture = () => screen.getByTitle(/Fixture/);
  const getTagName = () => screen.getByTitle(/Tag name/);
  const testPrevScroll = (value: number) => screen.getByText('Prev: ' + value);
  const testCurrScroll = (value: number) => screen.getByText('Curr: ' + value);
  const testValue = (value: number) => screen.getByText('Value: ' + value);
  const testResult = (result: ScrollResult) =>
    expect(screen.getByTitle('Result').innerHTML).toBe(result);

  const Fixture = (props?: Partial<ScrollConfig> & { isWindow?: boolean }) => {
    const [state, ref] = useScroll<HTMLDivElement>(props);

    return (
      <div
        ref={props?.isWindow ? undefined : ref}
        title="Fixture"
        style={{
          height: '300px',
          width: '300px',
          [props?.axis === 'x' ? 'overflowX' : 'overflowY']: 'scroll',
        }}
      >
        <span title="Tag name">{ref.current?.tagName}</span>
        <div title="Result" style={{ height: '500px', width: '500px' }}>
          {state.is}
        </div>
        {state.is !== 'idle' && (
          <>
            <div>Prev: {state.prev}</div>
            <div>Curr: {state.curr}</div>
            <div>Value: {state.value}</div>
          </>
        )}
      </div>
    );
  };

  it('[FRAGILE] the initial, no operation state is assigned', () => {
    const { result } = renderHook(() => useScroll());
    expect(result.current).toMatchSnapshot();
  });

  it('[FRAGILE] allows to assign reference to any HTML element', () => {
    render(<Fixture />);
    expect(getTagName().tagName).toBe('SPAN');
  });

  it('triggers scroll callback', async () => {
    const spy = jest.fn();

    render(<Fixture isWindow onScroll={spy} />);

    expect(window.scrollY).toBe(0);
    testResult('idle');

    fireEvent.scroll(window, { target: { scrollY: 120 } });

    await waitFor(() => {
      expect(window.scrollY).toBe(120);
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('window is used and state is changed during scrolling for', () => {
    const testScrollTracking = async (axis: ScrollAxis) => {
      const containerSizeKey = axis === 'y' ? 'innerHeight' : 'innerWidth';
      const scrollOffsetKey = axis === 'y' ? 'scrollY' : 'scrollX';
      const scrollSizeKey = axis === 'y' ? 'scrollHeight' : 'scrollWidth';
      const initScrollOffset = window[scrollOffsetKey];
      const initContainerSize = window[scrollOffsetKey];
      const initScrollSize = document.documentElement[scrollSizeKey];

      Object.defineProperty(window, scrollOffsetKey, { value: 0 });
      Object.defineProperty(window, containerSizeKey, { value: 970 });
      Object.defineProperty(document.documentElement, scrollSizeKey, {
        configurable: true,
        value: 2728,
      });

      render(<Fixture isWindow axis={axis} />);

      expect(window[scrollOffsetKey]).toBe(0);
      testResult('idle');

      fireEvent.scroll(window, { target: { [scrollOffsetKey]: 100 } });

      await waitFor(() => {
        expect(window[scrollOffsetKey]).toBe(100);
        testPrevScroll(0);
        testCurrScroll(100);
        testValue(5.69);
        testResult('progress');
      });

      fireEvent.scroll(window, { target: { [scrollOffsetKey]: 80 } });

      await waitFor(() => {
        expect(window[scrollOffsetKey]).toBe(80);
        testPrevScroll(100);
        testCurrScroll(80);
        testValue(4.55);
        testResult('regress');
      });

      fireEvent.scroll(window, { target: { [scrollOffsetKey]: 80 } });

      await waitFor(() => {
        expect(window[scrollOffsetKey]).toBe(80);
        testPrevScroll(80);
        testCurrScroll(80);
        testValue(4.55);
        testResult('unchanged');
      });

      Object.defineProperty(window, scrollOffsetKey, {
        value: initScrollOffset,
      });
      Object.defineProperty(window, containerSizeKey, {
        value: initContainerSize,
      });
      Object.defineProperty(document.documentElement, scrollSizeKey, {
        configurable: true,
        value: initScrollSize,
      });
    };

    it('x axis', async () => {
      await testScrollTracking('x');
    });

    it('y axios', async () => {
      await testScrollTracking('y');
    });
  });

  describe('reference element is used and state is changed during scrolling for', () => {
    const testScrollTracking = async (axis: ScrollAxis) => {
      const scrollOffsetKey = axis === 'y' ? 'scrollTop' : 'scrollLeft';

      render(<Fixture axis={axis} />);

      const element = getFixture();
      expect(element[scrollOffsetKey]).toBe(0);
      testResult('idle');

      fireEvent.scroll(element, { target: { [scrollOffsetKey]: 100 } });

      await waitFor(() => {
        expect(element[scrollOffsetKey]).toBe(100);
        testPrevScroll(0);
        testCurrScroll(100);
        testResult('progress');
      });

      fireEvent.scroll(element, { target: { [scrollOffsetKey]: 80 } });

      await waitFor(() => {
        expect(element[scrollOffsetKey]).toBe(80);
        testPrevScroll(100);
        testCurrScroll(80);
        testResult('regress');
      });

      fireEvent.scroll(element, { target: { [scrollOffsetKey]: 80 } });

      await waitFor(() => {
        expect(element[scrollOffsetKey]).toBe(80);
        testPrevScroll(80);
        testCurrScroll(80);
        testResult('unchanged');
      });
    };

    it('x axis', async () => {
      await testScrollTracking('x');
    });

    it('y axis', async () => {
      await testScrollTracking('y');
    });
  });
});
