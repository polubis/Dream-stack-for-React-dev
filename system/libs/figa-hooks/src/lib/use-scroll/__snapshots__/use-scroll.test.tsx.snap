// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll data detection works when: [FRAGILE] the initial, no operation state is assigned 1`] = `
Array [
  Object {
    "is": "idle",
  },
  Object {
    "current": null,
  },
]
`;
